File: top_interview_questions/2_medium/1_array_and_strings/longest_palindromic_substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        s = self.duplicateChars(s)
        longest = ""
        current = ""
        length = len(s)

        for n in range(0, length):
            i = 0
            while n - i >= 0 and n + i < length - 1:
                if not s[n - i] == s[n + i + 1]:
                    break
                i += 1

            current = s[n - i + 1:n + i + 1]
            longest = max(longest, current, key=len)

        return longest[::2]  # undo the duplicationg

    def duplicateChars(self, s: str) -> str:
        chars = []
        for char in s:
            chars.append(char)
            chars.append(char)

        return ''.join(chars)

File: top_interview_questions/2_medium/1_array_and_strings/longest_palindromic_substring/solution.go
package main

import "fmt"

func main() {
	fmt.Println(longestPalindrome("babad")) // bab
	fmt.Println(longestPalindrome("cbbd"))  // bb
}

func longestPalindrome(s string) string {
	s = duplicateChars(s)
	longest := ""
	current := ""
	length := len(s)

	for n := 0; n < length; n++ {
		i := 0
		for n-i >= 0 && n+i < length-1 {
			fmt.Println(n, i)
			if s[n-i] != s[n+i+1] {
				break
			}
			i++
		}

		current = s[n-i+1 : n+i+1]
		if len(current) > len(longest) {
			longest = current
		}
	}

	return removeDuplicates(longest)
}

func duplicateChars(s string) string {
	r := ""

	for _, char := range s {
		r += string(char)
		r += string(char)
	}

	return r
}

func removeDuplicates(s string) string {
	r := ""
	for i, char := range s {
		if i%2 == 0 {
			r += string(char)
		}
	}

	return r
}
